Mathlib/Lean/Expr/Basic.lean

```diff
@@ -187,6 +187,17 @@ def getAppApps (e : Expr) : Array Expr :=
   let nargs := e.getAppNumArgs
   getAppAppsAux e (mkArray nargs dummy) (nargs-1)
 
+/-- Erase proofs in an expression by replacing them with `sorry`s.
+
+Useful, e.g., to verify if the proof-irrelevant part of a definition depends on a variable. -/
+def eraseProofs (e : Expr) : MetaM Expr :=
+  Meta.transform e
+    (pre := fun e => do
+      if (← Meta.isProof e) then
+        return .done (← mkSyntheticSorry (← inferType e))
```

Comment:
Note that the expression may still contain proofs after `eraseProofs`, as there might be proofs in the type of a proof within `e` (and this winds up in `e` proper, as an argument to `sorryAx`).

I'm struggling to think of any practical (non-contrived) examples, however. Plus, the current implementation has the advantage of preserving the type. Do you think this could ever be an issue in the application at hand?

If you wanted to really make sure there were no proofs in the expression, you could write
```suggestion
partial def eraseProofs (e : Expr) : MetaM Expr :=
  Meta.transform e
    (pre := fun e => do
      if (← Meta.isProof e) then
        return .done (← mkSyntheticSorry <|← (← inferType e).eraseProofs)
```
but at the cost of potentially not preserving the type. Otherwise, it might simply be worth adding a caveat in the docstring.

Reply:
@kmill What do you think? Should we have 2 functions?

Reply:
Are you sure that this doesn't preserve the type @thorimur ? Proof irrelevance should mean that the type is the same, up to defeq.

For efficiency (and incidentally avoiding `partial`), I would instead change that line to
```
return .continue (← mkSyntheticSorry (← inferType e))
```

Reply:
True, I'm talking about preservation on the nose (`BEq`) and should have specified. (I thought it was nice that (as far as I could see) all mdata and the like was preserved.) Defeq is probably all we really need.

Ah, I see, `.continue` makes more sense there!

Reply:
I tried to follow suggestions in this thread. Please have another look.